plugins/net_plugin/net_plugin.cpp

```diff
@@ -3428,10 +3470,53 @@ namespace eosio {
       }
    }
 
+   // called from any thread
+   void net_plugin_impl::update_security_group(const block_state_ptr& bs) {
+      // update cache
+      //    - connection_mtx is needed
+      //
+      auto& update = bs->get_security_group_info();
+      std::shared_lock<std::shared_mutex> connection_guard(connections_mtx);
+      if(!security_group.update_cache(update.version, update.participants)) {
+         return;
+      }
+
+      std::vector<connection_ptr> added_connections;
+      added_connections.reserve(connections.size());
+
+      // update connections
+      //    - connection_mtx still needed
+      //
+      for(auto& connection : connections) {
+         const auto& participant = connection->participant_name();
+         if(!participant) {
+            return;
+         }
+         if(security_group.is_in_security_group(participant.value())) {
+            if(!connection->is_participating()) {
+               connection->set_participating(true);
+               added_connections.push_back(connection);
+            }
+         }
+         else {
+            connection->set_participating(false);
+         }
+      };
+
+      // send handshake when added to group
+      //    - connection mutex no longer needd
+      //
+      connection_guard.unlock();
+      for(auto& connection : added_connections) {
+         connection->send_handshake();
```

Comment:
This just does a post on a strand. I don't see the need to pull this out and send after releasing the mutex. Just call send_handshake() above so that `added_connections` is not needed.

Reply:
Removed added_connections and now call send_handsahke in update loop